File: ffmpeg-javacpp/src/main/kotlin/com/leovp/ffmpeg/javacpp/audio/adpcm/AdpcmImaQTDecoder.kt
@file:Suppress("unused")

package com.leovp.ffmpeg.javacpp.audio.adpcm

import org.bytedeco.ffmpeg.avcodec.AVCodecContext
import org.bytedeco.ffmpeg.avcodec.AVPacket
import org.bytedeco.ffmpeg.avutil.AVDictionary
import org.bytedeco.ffmpeg.avutil.AVFrame
import org.bytedeco.ffmpeg.global.avcodec
import org.bytedeco.ffmpeg.global.avutil
import org.bytedeco.javacpp.BytePointer
import java.io.Closeable

class AdpcmImaQTDecoder(sampleRate: Int, private val channel: Int) : Closeable {
    private var context: AVCodecContext? = null
    private var packet: AVPacket? = null
    private var frame: AVFrame? = null

    init {
        require(sampleRate > 0) { "Sample rate must be positive." }
        require(channel == 1 || channel == 2) { "Channel count must be 1 or 2." }
        initialize(sampleRate)
    }

    private fun initialize(sampleRate: Int) {
        try {
            val codec = avcodec.avcodec_find_decoder(avcodec.AV_CODEC_ID_ADPCM_IMA_QT)
            check(codec != null && !codec.isNull) { "ADPCM IMA QT decoder was not found." }
            context = avcodec.avcodec_alloc_context3(codec)
            val initializedContext = context
            check(initializedContext != null && !initializedContext.isNull) {
                "Unable to allocate the ADPCM codec context."
            }
            initializedContext.ch_layout(
                if (channel == 1) {
                    avutil.AV_CHANNEL_LAYOUT_MONO
                } else {
                    avutil.AV_CHANNEL_LAYOUT_STEREO
                }
            )
            initializedContext.sample_rate(sampleRate)
            check(avcodec.avcodec_open2(initializedContext, codec, null as AVDictionary?) >= 0) {
                "Unable to open the ADPCM codec."
            }

            packet = avcodec.av_packet_alloc()
            check(packet != null && packet?.isNull == false) { "Unable to allocate an AVPacket." }
            frame = avutil.av_frame_alloc()
            check(frame != null && frame?.isNull == false) { "Unable to allocate an AVFrame." }
        } catch (throwable: Throwable) {
            releaseResources()
            throw throwable
        }
    }

    /** In QuickTime, one IMA chunk contains 34 bytes and decodes to 64 samples per channel. */
    fun chunkSize(): Int = ENCODED_CHUNKS_SIZE * channel

    /**
     * Decodes one ADPCM IMA QT chunk into separate PCM channel buffers.
     *
     * Mono output is returned as `leftBytes to ByteArray(0)`. A codec rejection returns `null`.
     */
    @Synchronized
    fun decode(adpcmBytes: ByteArray): Pair<ByteArray, ByteArray>? {
        require(adpcmBytes.size == chunkSize()) {
            "Invalid ChunkSize: ${adpcmBytes.size}, required: ${chunkSize()}, " +
                "In QuickTime, IMA is encoded by chunks of 34*channels bytes (=64 samples)"
        }
        val initializedContext = checkNotNull(context) { "Decoder is closed." }
        val reusablePacket = checkNotNull(packet) { "Decoder is closed." }
        val reusableFrame = checkNotNull(frame) { "Decoder is closed." }

        avcodec.av_packet_unref(reusablePacket)
        avutil.av_frame_unref(reusableFrame)
        check(avcodec.av_new_packet(reusablePacket, adpcmBytes.size) >= 0) {
            "Unable to allocate the ADPCM packet payload."
        }
        try {
            val packetData = reusablePacket.data()
            check(packetData != null && !packetData.isNull) { "ADPCM packet has no payload." }
            packetData.position(0).put(adpcmBytes, 0, adpcmBytes.size)
            packetData.position(adpcmBytes.size.toLong()).put(INPUT_PADDING, 0, INPUT_PADDING.size)
            packetData.position(0)

            if (
                avcodec.avcodec_send_packet(initializedContext, reusablePacket) < 0 ||
                avcodec.avcodec_receive_frame(initializedContext, reusableFrame) < 0
            ) {
                return null
            }
            return copyDecodedChannels(reusableFrame)
        } finally {
            avutil.av_frame_unref(reusableFrame)
            avcodec.av_packet_unref(reusablePacket)
        }
    }

    private fun copyDecodedChannels(decodedFrame: AVFrame): Pair<ByteArray, ByteArray> {
        val bytesPerSample = avutil.av_get_bytes_per_sample(decodedFrame.format())
        check(bytesPerSample > 0 && decodedFrame.nb_samples() > 0) {
            "Decoder returned an invalid sample format or sample count."
        }
        val decodedChannels = decodedFrame.ch_layout().nb_channels()
        check(decodedChannels == channel) { "Decoder returned an unexpected channel count." }
        val bytesPerChannel = Math.multiplyExact(decodedFrame.nb_samples(), bytesPerSample)
        val planar = avutil.av_sample_fmt_is_planar(decodedFrame.format()) == 1

        if (planar) {
            val left = readBytes(decodedFrame.extended_data(0), bytesPerChannel)
            val right = if (decodedChannels == 2) {
                readBytes(decodedFrame.extended_data(1), bytesPerChannel)
            } else {
                ByteArray(0)
            }
            return left to right
        }

        val packedSize = Math.multiplyExact(bytesPerChannel, decodedChannels)
        val packed = readBytes(decodedFrame.extended_data(0), packedSize)
        if (decodedChannels == 1) return packed to ByteArray(0)

        val left = ByteArray(bytesPerChannel)
        val right = ByteArray(bytesPerChannel)
        repeat(decodedFrame.nb_samples()) { sampleIndex ->
            val packedOffset = sampleIndex * decodedChannels * bytesPerSample
            val channelOffset = sampleIndex * bytesPerSample
            packed.copyInto(left, channelOffset, packedOffset, packedOffset + bytesPerSample)
            packed.copyInto(
                right,
                channelOffset,
                packedOffset + bytesPerSample,
                packedOffset + bytesPerSample * 2
            )
        }
        return left to right
    }

    private fun readBytes(pointer: BytePointer?, length: Int): ByteArray {
        check(pointer != null && !pointer.isNull) { "Decoder returned a null audio plane." }
        return ByteArray(length).also { pointer.position(0).get(it) }
    }

    @Synchronized
    override fun close() {
        releaseResources()
    }

    private fun releaseResources() {
        frame?.let(avutil::av_frame_free)
        frame = null
        packet?.let(avcodec::av_packet_free)
        packet = null
        context?.let(avcodec::avcodec_free_context)
        context = null
    }

    companion object {
        private const val ENCODED_CHUNKS_SIZE = 34
        private val INPUT_PADDING = ByteArray(avcodec.AV_INPUT_BUFFER_PADDING_SIZE)

        fun getSampleFormatName(fmt: Int): String? = avutil.av_get_sample_fmt_name(fmt).let { ptr ->
            return if (ptr != null && !ptr.isNull) ptr.string else null
        }

        fun isAvSampleInPlanar(fmt: Int): Boolean = avutil.av_sample_fmt_is_planar(fmt) == 1
    }
}

File: ffmpeg-sdk/src/main/cpp/adpcm_ima_qt_decoder/adpcm_ima_qt_decoder.cpp
#include "adpcm_ima_qt_decoder.h"
#include "logger.h"

#include <cstring>
#include <limits>
#include <new>

AdpcmImaQtDecoder::AdpcmImaQtDecoder(int sampleRate, int channels) {
    LOGE("ADPCM decoder init. sampleRate: %d, channels: %d", sampleRate, channels);

    this->sampleRate = sampleRate;
    this->channels = channels;

    const AVCodec *codec = avcodec_find_decoder(AV_CODEC_ID_ADPCM_IMA_QT);
    if (!codec) {
        LOGE("Decoder: ADPCM IMA QT decoder not found");
        return;
    }
    ctx = avcodec_alloc_context3(codec);
    if (!ctx) {
        LOGE("Decoder: Could not allocate codec context");
        return;
    }
    ctx->sample_rate = sampleRate;
    ctx->ch_layout = channels == 2 ? (AVChannelLayout) AV_CHANNEL_LAYOUT_STEREO : (AVChannelLayout) AV_CHANNEL_LAYOUT_MONO;

    int ret = avcodec_open2(ctx, codec, nullptr);
    if (ret < 0) {
        LOGE("Decoder: avcodec_open2 error. code=%d", ret);
        avcodec_free_context(&ctx);
        ctx = nullptr;
        return;
    }

    frame = av_frame_alloc();
    pkt = av_packet_alloc();
    if (frame == nullptr || pkt == nullptr) {
        LOGE("Decoder: Could not allocate frame or packet");
        if (frame != nullptr) av_frame_free(&frame);
        if (pkt != nullptr) av_packet_free(&pkt);
        avcodec_free_context(&ctx);
        return;
    }
    valid = true;
}

AdpcmImaQtDecoder::~AdpcmImaQtDecoder() {
    if (frame != nullptr) {
        av_frame_free(&frame);
        frame = nullptr;
    }
    if (pkt != nullptr) {
        av_packet_free(&pkt);
        pkt = nullptr;
    }
    if (ctx != nullptr) {
        avcodec_free_context(&ctx);
        ctx = nullptr;
    }
    LOGE("ADPCM decoder released!");
}

uint8_t *AdpcmImaQtDecoder::decode(const uint8_t *adpcmByteArray, int adpcmLength,
                                   int *outPcmLength) {
    *outPcmLength = 0;
    av_packet_unref(pkt);
    av_frame_unref(frame);
    int ret = av_new_packet(pkt, adpcmLength);
    if (ret < 0) {
        LOGE("Decoder: av_new_packet() error. code=%d", ret);
        return nullptr;
    }
    memcpy(pkt->data, adpcmByteArray, adpcmLength);
    memset(pkt->data + adpcmLength, 0, AV_INPUT_BUFFER_PADDING_SIZE);
    if ((ret = avcodec_send_packet(ctx, pkt)) < 0) {
        LOGE("Decoder: avcodec_send_packet() error. code=%d", ret);
        av_packet_unref(pkt);
        return nullptr;
    }
    if ((ret = avcodec_receive_frame(ctx, frame)) < 0) {
        LOGE("Decoder: avcodec_receive_frame() error. code=%d", ret);
        av_packet_unref(pkt);
        return nullptr;
    }

    const int bytes_per_sample = av_get_bytes_per_sample((AVSampleFormat) frame->format);
    const int channel_count = frame->ch_layout.nb_channels;
    const int64_t each_channel_length =
            static_cast<int64_t>(frame->nb_samples) * bytes_per_sample;
    const int64_t pcm_size = each_channel_length * channel_count;
    if (bytes_per_sample <= 0 || frame->nb_samples <= 0 ||
        channel_count != channels || each_channel_length > std::numeric_limits<int>::max() ||
        pcm_size <= 0 || pcm_size > std::numeric_limits<int>::max()) {
        av_packet_unref(pkt);
        av_frame_unref(frame);
        return nullptr;
    }
    const int pcmSize = static_cast<int>(pcm_size);
    *outPcmLength = pcmSize;

    auto *outPcmBytes = new(std::nothrow) uint8_t[pcmSize];
    if (outPcmBytes == nullptr) {
        av_packet_unref(pkt);
        av_frame_unref(frame);
        *outPcmLength = 0;
        return nullptr;
    }
    if (av_sample_fmt_is_planar((AVSampleFormat) frame->format) && channel_count > 1) {
        if (frame->extended_data == nullptr) {
            delete[] outPcmBytes;
            av_packet_unref(pkt);
            av_frame_unref(frame);
            *outPcmLength = 0;
            return nullptr;
        }
        for (int channel = 0; channel < channel_count; ++channel) {
            if (frame->extended_data[channel] == nullptr) {
                delete[] outPcmBytes;
                av_packet_unref(pkt);
                av_frame_unref(frame);
                *outPcmLength = 0;
                return nullptr;
            }
        }
        for (int sample = 0; sample < frame->nb_samples; ++sample) {
            for (int channel = 0; channel < channel_count; ++channel) {
                memcpy(outPcmBytes + (sample * channel_count + channel) * bytes_per_sample,
                       frame->extended_data[channel] + sample * bytes_per_sample,
                       bytes_per_sample);
            }
        }
    } else {
        if (frame->data[0] == nullptr) {
            delete[] outPcmBytes;
            av_packet_unref(pkt);
            av_frame_unref(frame);
            *outPcmLength = 0;
            return nullptr;
        }
        memcpy(outPcmBytes, frame->data[0], pcmSize);
    }
    av_packet_unref(pkt);
    av_frame_unref(frame);
    return outPcmBytes;
}

AVCodecContext *AdpcmImaQtDecoder::getCodecContext() {
    return ctx;
}

[[maybe_unused]] int AdpcmImaQtDecoder::getSampleRate() const {
    return sampleRate;
}

int AdpcmImaQtDecoder::getChannels() const {
    return channels;
}
